src/main/java/org/moltimate/moltimatebackend/service/AlignmentService.java

```diff
@@ -234,4 +264,100 @@ private double rmsd(String motifId, List<Residue> activeSiteResidues, List<Group
         Point3d[] point3ds = new Point3d[points.size()];
         return points.toArray(point3ds);
     }
+
+    /**
+     * Finds all permutations of a residue mapping
+     *
+     * @param residueMapping: mapping of residue to its candidate matches for alignment
+     * @return a list of mappings from residue to group where each map is a possible alignment permutation
+     */
+    private List<Map<Residue, Group>> findAllPermutations(Map<Residue, List<Group>> residueMapping) {
+        List<Map<Residue, Group>> result = new ArrayList<>();
+        List<List<Group>> groupLists = new ArrayList<>(residueMapping.values());
+        List<List<Group>> results = cartesianProduct(groupLists);
+        List<Residue> residues = new ArrayList<>(residueMapping.keySet());
+
+        for (List<Group> res : results) {
+            Map<Residue, Group> mapping = new HashMap<>();
+            for (int i = 0; i < res.size(); i++) {
+                mapping.put(residues.get(i), res.get(i));
+            }
+            result.add(mapping);
+        }
+        return result;
+    }
+
+    /**
+     * Finds the best permutation of a list of permutations of matches to an active site
+     * of a motif
+     *
+     * @param motifStruct:  structure object of motif to find permutation of
+     * @param permutations: data structure containing your permutations
+     * @param motif:        motif that the alignment permutations relate to
+     * @return best fit permutation for alignment. We calculate this by checking if the permutation fits
+     * the constraint we have for levenstein distance and then find the one with minimum RMSD that
+     * fits this requirement.
+     */
+    private Map<Residue, Group> findBestPermutation(Structure motifStruct, List<Map<Residue, Group>> permutations, Motif motif) {
+        double min_rmsd = Double.MAX_VALUE;
+        Map<Residue, Group> best_match = new HashMap<>();
+        for (Map<Residue, Group> permutation : permutations) {
+            List<Group> alignmentSeq = new ArrayList<>(permutation.values());
+            alignmentSeq.sort(Comparator.comparingInt(o -> o.getResidueNumber()
+                    .getSeqNum()));
+
+            String alignmentString = AlignmentUtils.groupListToResString(alignmentSeq);
+            String motifResString = AlignmentUtils.residueListToResString(motif.getActiveSiteResidues());
+
+            int distance = AlignmentUtils.levensteinDistance(alignmentString, motifResString);
+
+            if (acceptableDistance(motif.getActiveSiteResidues()
+                                           .size(), distance)) {
+
+                double rmsd = rmsd(motifStruct, motif.getActiveSiteResidues(), alignmentSeq);
```

Comment:
This is incorrect. It's calculating the RMSD of the possible alignment so we can choose the best one. We need to do this in order to choose the best alignment from the set of possible candidates.

Reply:
_Spoke in person_

This is the RMSD of the possible alignment. I asked the follow-up question, "Can we use this result later on so we don't have to recalculate RMSD?"
Answer: We could do that but it would be messy to manage the primary case (where this code isn't used since we didn't have multiple matches to a search query).
IMPACT: No changes needed